lib/interleaved_tests.rb

```diff
@@ -0,0 +1,286 @@
+      end
+
+      @all_examples << example
+      @pending_examples << example
+    end
+
+    def example_failed(example)
+      notification = example.notification
+
+      @formatters.each do |formatter|
+        formatter.example_failed(notification) if formatter.respond_to?(:example_failed)
+      end
+
+      @all_examples << example
+      @failed_examples << example
+    end
+
+    def method_missing(method, arg)
+      @formatters.each do |formatter|
+        formatter.send(method, arg) if formatter.respond_to?(method)
+      end
+    end
+  end
+
+  FakeException = Struct.new(:backtrace, :message, :cause)
+  class FakeException
+    def self.from_obj(obj)
+      if obj
+        obj = obj.symbolize_keys
+        new(
+          obj[:backtrace],
+          obj[:message],
+          obj[:cause]
+        )
+      end
+    end
+  end
+
+  FakeExecutionResult = Struct.new(:example_skipped?, :pending_message, :status, :pending_fixed?, :exception)
+  class FakeExecutionResult
+    def self.from_obj(obj)
+      obj = obj.symbolize_keys
+      new(
+        obj[:example_skipped?],
+        obj[:pending_message],
+        obj[:status].to_sym,
+        obj[:pending_fixed?],
+        FakeException.from_obj(obj[:exception])
+      )
+    end
+  end
+
+  FakeExample = Struct.new(:execution_result, :location, :full_description, :metadata, :location_rerun_argument)
+  class FakeExample
+    def self.from_obj(obj)
+      obj = obj.symbolize_keys
+      new(
+        FakeExecutionResult.from_obj(obj[:execution_result]),
+        obj[:location],
+        obj[:full_description],
+        obj[:metadata].symbolize_keys,
+        obj[:location_rerun_argument],
+      )
+    end
+
+    def notification
+      RSpec::Core::Notifications::ExampleNotification.for(
+        self
+      )
+    end
+  end
+
+  def self.run(formatter_config, files)
+    start_time = Time.now
+
+    reporter = Reporter.from_config(formatter_config)
+
+    num_processes = ParallelTests.determine_number_of_processes(nil)
+
+    tests_in_groups =
+      ParallelTests::RSpec::Runner.tests_in_groups(
+        files,
+        num_processes,
+        group_by: :filesize
+      )
+
+    messages = Queue.new
+
+    `rm -r tmp/test-pipes`
+    `mkdir -p tmp/test-pipes/`
+
+    tests_in_groups.each_with_index do |tests, process_num|
+      process_num += 1
+
+      if tests.empty?
+        messages << {type: 'exit', process_num: process_num}
+      else
+        begin
+          File.mkfifo("tmp/test-pipes/subprocess-#{process_num}")
+        rescue Errno::EEXIST
+        end
+
+        command = [
+          "TEST_ENV_NUMBER=#{process_num}",
+          "bundle exec rspec",
+          "-f JsonRowsFormatter",
+          "-o tmp/test-pipes/subprocess-#{process_num}",
+          tests.map(&:inspect).join(' ')
+        ].join(' ')
+
+        stdin, stdout, stderr, wait_thr = Open3.popen3(command)
+
+        Thread.new do
```

Comment:
I worry a bit about having a bunch of loose threads that are never joined to, stored in instance vars or guaranteed to be terminated in an orderly fashion, should you be storing them in instance vars and checking on them?